src/Stream.re

```diff
@@ -1,74 +1,53 @@
-type sendFunc('a) = 'a => unit;
-/* type complete = unit => unit; */
-
-type streamFunc('a) = sendFunc('a) => unit;
-
-type unsubscribeFunc = unit => unit;
-
-type subscriber('a) = {
+type subscriber('msg) = {
   id: int,
-  onValue: sendFunc('a),
+  onValue: 'msg => unit,
   /* onComplete: complete, */
 };
 
-type t('a) = {
+type t('msg) = {
   lastSubscriberId: ref(int),
-  subscribers: ref(list(subscriber('a))),
+  subscribers: ref(list(subscriber('msg))),
 };
 
-let ofDispatch = (f: streamFunc('a)) => {
-  let subscribers: ref(list(subscriber('a))) = ref([]);
+let ofDispatch = f => {
+  let subscribers: ref(list(subscriber('msg))) = ref([]);
 
-  let onValue = (v: 'a) => {
-    List.iter(s => s.onValue(v), subscribers^);
-  };
+  let onValue = value => List.iter(sub => sub.onValue(value), subscribers^);
 
   f(onValue);
 
   {lastSubscriberId: ref(0), subscribers};
 };
 
 let create = () => {
-  let _dispatchFunction = ref(None);
+  let dispatcher = ref(None);
 
-  let stream = ofDispatch(dispatch => _dispatchFunction := Some(dispatch));
+  let stream = ofDispatch(dispatch => dispatcher := Some(dispatch));
 
-  let dispatch = v => {
-    switch (_dispatchFunction^) {
-    | Some(f) => f(v)
-    | None => ()
-    };
-  };
+  let dispatch = msg => Option.iter(dispatch => dispatch(msg), dispatcher^);
 
   (stream, dispatch);
```

Comment:
This thing is really complicated, and I don't understand why. I also don't dare touch it since it's not tested, but I don't see why this complexity should be necessary when you have a list of subscribers.

Reply:
For sure... I'd like to just remove this complexity.